Add blog post about LoRa fine-tuning for sequence classification

mehdiir opened the pull request:
The main objective of this blogpost is to implement LoRa fine-tuning for sequence classification task using three pre-trained models from HuggingFace: [meta-llama/Llama-2-7b-hf](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf), [mistralai/Mistral-7B-v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1), and [roberta-large](https://huggingface.co/roberta-large)

mehdiir commented:
Link to PR on the HF documentation-images repo: https://huggingface.co/datasets/huggingface/documentation-images/discussions/203

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+    - [LoRA](#lora)
+- [Setup](#setup)
+- [Data preparation](#data-preparation)
+    - [A. Data loading](#a-data-loading)
```
instead of having numbered list in outline you can just do indents IMHO.

mehdiir replied:
Done. I removed the numbers.

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+<!-- /TOC -->
+# Introduction 
+
+In the fast-moving world of Natural Language Processing (NLP), we often find ourselves comparing different language models to see which one works best for specific tasks. This blog post is all about comparing three models: RoBERTa, Mistral-7b, and Llama-2-7b. We used them to tackle a common problem - classifying tweets about disasters. It is important to note that Mistral and Llama are large mode ...
```
```suggestion
In the fast-moving world of Natural Language Processing (NLP), we often find ourselves comparing different language models to see which one works best for specific tasks. This blog post is all about comparing three models: RoBERTa, Mistral-7b, and Llama-2-7b. We used them to tackle a common problem - classifying tweets about disasters. It is important to note that Mistral and Llama are large models with 7 billion parameters. In contrast, RoBERTa-large (355M parameters) is a relatively smaller model used as a baseline for the comparison study.
```
Also maybe have a small note that they're essentially different type of models, i.e. one is used for understanding and other two are for generation (so it's a fair comparison) (because it's known that BERT-like models work better in these tasks in small scale)

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+In the fast-moving world of Natural Language Processing (NLP), we often find ourselves comparing different language models to see which one works best for specific tasks. This blog post is all about comparing three models: RoBERTa, Mistral-7b, and Llama-2-7b. We used them to tackle a common problem - classifying tweets about disasters. It is important to note that Mistral and Llama are large mode ...
+
+In this blog, we used PEFT (Parameter-Efficient Fine-Tuning) technique: Lora (Low-Rank Adaptation of Large Language Models) for fine-tuning the pre-trained model on a sequence classification task. LoRa is designed to significantly reduce the number of trainable parameters while maintaning strong downstream task performance. 
```
```suggestion
In this blog, we used PEFT (Parameter-Efficient Fine-Tuning) technique: LoRA (Low-Rank Adaptation of Large Language Models) for fine-tuning the pre-trained model on the sequence classification task. LoRa is designed to significantly reduce the number of trainable parameters while maintaining strong downstream task performance. 
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+In this blog, we used PEFT (Parameter-Efficient Fine-Tuning) technique: Lora (Low-Rank Adaptation of Large Language Models) for fine-tuning the pre-trained model on a sequence classification task. LoRa is designed to significantly reduce the number of trainable parameters while maintaning strong downstream task performance. 
+
+The main objective of this blogpost is to implement LoRa fine-tuning for sequence classification task using three pre-trained models from HuggingFace: [meta-llama/Llama-2-7b-hf](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf), [mistralai/Mistral-7B-v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1), and [roberta-large](https://huggingface.co/roberta-large)
```
```suggestion
The main objective of this blogpost is to implement LoRA fine-tuning for sequence classification task using three pre-trained models from Hugging Face: [meta-llama/Llama-2-7b-hf](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf), [mistralai/Mistral-7B-v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1), and [roberta-large](https://huggingface.co/roberta-large)
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+The main objective of this blogpost is to implement LoRa fine-tuning for sequence classification task using three pre-trained models from HuggingFace: [meta-llama/Llama-2-7b-hf](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf), [mistralai/Mistral-7B-v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1), and [roberta-large](https://huggingface.co/roberta-large)
+
+# Hardware used 
```
```suggestion
# Hardware Used 
```
Maybe have some explanation, e.g. We used an A6000 with following specs

pcuenca replied:
```suggestion
## Hardware Used
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+# Learnings 
+
+- Implement fine-tuning of pre-trained LLMs using LoRa PEFT methods (Lora).
```
```suggestion
- Implement fine-tuning of pre-trained LLMs using LoRA PEFT methods.
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+- Implement fine-tuning of pre-trained LLMs using LoRa PEFT methods (Lora).
+- Learn how to use the HuggingFace APIs ([transformers](https://huggingface.co/docs/transformers/index), [peft](https://huggingface.co/docs/peft/index), and [datasets](https://huggingface.co/docs/datasets/index)).
+- Setup the hyperparameters tuning and experiment logging using weights & biaises ([wandb](https://wandb.ai)).
```
```suggestion
- Setup the hyperparameter tuning and experiment logging using [Weights & Biases](https://wandb.ai)).
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+
+
+# Pre-requisites 
```
Again, maybe add an explanation, also they're dependencies (when we say pre-requisites it sounds different)

mehdiir replied:
Modified to dependencies.

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,798 @@
+memory usage: 127.6+ KB
+```
+
+- Target distribution in the train dataset
````
```suggestion
**Target distribution in the train dataset**
```
Having these in bold would look better IMHO

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+### A. Load checkpoints for the classfication model  
+
+We load the pre-trained RoBERTa model with a sequence classification header using the HuggingFace `AutoModelForSequenceClassification` class:
```
```suggestion
We load the pre-trained RoBERTa model with a sequence classification head using the Hugging Face `AutoModelForSequenceClassification` class:
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+In the code below, we use the values recommended by the [Lora paper](https://arxiv.org/abs/2106.09685).
+
+Note that in this blog, we will perform a hyperparameters tunning with wandb.
```
```suggestion
Note that in this blog, we will perform a hyperparameter tuning with wandb.
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+### A. Load checkpoints for the classfication model
+
+ Let's load pre-trained Mistral 7B model with a sequence classification header:
```
```suggestion
Let's load the pre-trained Mistral-7B model with a sequence classification head:
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+### B.  LoRa setup for Mistral 7B classifier
+
+For Mistral 7B model, we need to specify the target_modules (The query and value vectors from the attention modules):
```
```suggestion
For Mistral 7B model, we need to specify the `target_modules` (the query and value vectors from the attention modules):
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+## Evaluation Metrics
+
+First, we define the perfomance metrics we will use to compare the three models: F1 score, recall, precision and accuracy:
```
```suggestion
First, we define the performance metrics we will use to compare the three models: F1 score, recall, precision and accuracy:
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,798 @@
+```
+
+## Custom Trainer for Weighted Loss 
+As mentioned at the begining of this post, we have an imbalanced distribution between positive and negative classes. To account for that, we need to train our models with a weight cross-entropy loss. The `Trainer` class doesn't support providing a custom loss as it expects getting the loss directly from the model's outputs. 
````
```suggestion
As mentioned at the beginning of this post, we have an imbalanced distribution between positive and negative classes. We need to train our models with a weight cross-entropy loss to account for that. The `Trainer` class doesn't support providing a custom loss as it expects to get the loss directly from the model's outputs. 
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,798 @@
+Then, we set the training arguments:
+
+```python
+from transformers import TrainingArguments, Trainer
````
Why import `Trainer` here as you're not using it?

mehdiir replied:
I removed it.

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+Let's set the training arguments and the trainer for the three models.
+
+### A. RoBERTa 
+First important step is to move the models to the GPU device for training
```
```suggestion
First important step is to move the models to the GPU device for training.
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,798 @@
+)
+```
+
+### B. Mistral 7B
````
```suggestion
### Mistral-7B
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+### B. Mistral 7B
+
+Similar to RoBERTa, we define the trainer as follows: 
```
```suggestion
Similar to RoBERTa, we initialize the `WeightedBCELossTrainer` as follows: 
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,798 @@
+)
+```
+
+**Note** that we needed to enable half-precision training by setting `fp16` to `True`. The main reason is that Mistral 7B is large and its weights cannot fit into one GPU memory (48GB) with full float32 precision. 
````
```suggestion
**Note** that we needed to enable half-precision training by setting `fp16` to `True`. The main reason is that Mistral-7B is large, and its weights cannot fit into one GPU memory (48GB) with full float32 precision. 
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,798 @@
+```python
+from transformers import TrainingArguments, Trainer
+
+mistral_model = mistral_model.cuda()
````
Do you need to do this again? Did you meant to take LLaMA to GPU instead?

mehdiir replied:
I modified it to Llama.

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+def llama_preprocessing_function(examples):
+    return llama_tokenizer(examples['text'], truncation=True, max_length=MAX_LEN)
+
+llama_tokenized_datasets = data.map(llama_preprocessing_function, batched=False)
```
Completely up to you: since you did the same thing above, you could refactor this part in one preprocessing function and only pass the dataset and tokenizer there instead to make it tidier.

mehdiir replied:
I think it is more easier to understand as it is.

mehdiir replied:
One clarification regarding my earlier point: I wrote standalone code for each model, allowing users interested in a specific model to follow the relevant section only.

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+
+
+# Hyper parameters tunning
```
```suggestion
# Hyperparameter Tuning
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+# Hyper parameters tunning
+
+We have used Wandb sweeps API to run hyperparameter tunning with a bayesian search strategy (30 runs). The hyper parameters tunned are :
```
```suggestion
We have used Wandb Sweep API to run hyperparameter tunning with Bayesian search strategy (30 runs). The hyperparameters tuned are the following.
```
Maybe you could give a link to Sweep API.

merveenoyan reviewed (commented):
Very interesting findings, I was wondering how larger autoregressive models would compare with encoder only models. I left a couple of comments about the code and did nitpicking, thanks a lot for this!

merveenoyan commented:
Also, if you wish, we could have our designer design a blog banner @mehdiir 😊

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+- GPU memory: 48GB x2
+
+
+# Learnings 
```
```suggestion
## Goals
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+
+
+# Pre-requisites 
```
```suggestion
## Dependencies
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,798 @@
+```
+
+
+# Pre-trained Models
+## [RoBERTa](https://arxiv.org/abs/1907.11692)
````
```suggestion
## Pre-trained Models

### [RoBERTa](https://arxiv.org/abs/1907.11692)
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+RoBERTa (Robustly Optimized BERT Approach) is an advanced variant of the BERT model proposed by Meta AI research team. It's a transformer-based language model using self-attention mechanisms for contextual word representations and trained with masked language model.
+
+The main differences of RoBERTa over BERT are:  
+- No Next Sentence Prediction (NSP): RoBERTa removes NSP pre-training task, which predicts if document segments come from the same document. Authors show that removing this step has a little impact on downstream tasks and in some cases it even improves the performances.
+- Larger Batches, Longer Sequences: While BERT's pre-training consists of 1 million steps with a batch size of 256 sequences, RoBERTa is trained with 125 steps involving 2,000 sequences, followed by 31,000 steps using batch sizes of 8,000 sequences.
+- Dynamic Masking: Unlike BERT's static masks, RoBERTa applies dynamic masks by generating new masked sequences at each training pass (epoch). This approach improves the performance across tasks as it can be seen as a data augmentation helping the model to learn from diverse contexts.
```
This is very cool, but in the interest of brevity maybe it suffices to say that RoBERTa is a popular model to fine-tune and appropriate as a baseline for our experiments. Again, we can link to the docs or the model card for people interested in learning more.

mehdiir replied:
Done.

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+- Dynamic Masking: Unlike BERT's static masks, RoBERTa applies dynamic masks by generating new masked sequences at each training pass (epoch). This approach improves the performance across tasks as it can be seen as a data augmentation helping the model to learn from diverse contexts.
+
+
+## [Llama 2](https://arxiv.org/abs/2307.09288)
```
```suggestion
### [Llama 2](https://arxiv.org/abs/2307.09288)
```

mehdiir replied:
Done.

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+Llama models, which stands for Large Language Model Meta AI, belong to the family of large language models (LLMs) introduced by Meta AI. The Llama models vary in size, with parameter counts ranging from 7 billion to 65 billion.
+
+Llama model is an auto-regressive language model, based on the transformer decoder architecture. To generate text, Llama processes a sequence of words as input and iteratively predicts the next token using a sliding window.
```
```suggestion
Llama is an auto-regressive language model, based on the transformer decoder architecture. To generate text, Llama processes a sequence of words as input and iteratively predicts the next token using a sliding window.
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+Llama models, which stands for Large Language Model Meta AI, belong to the family of large language models (LLMs) introduced by Meta AI. The Llama models vary in size, with parameter counts ranging from 7 billion to 65 billion.
+
+Llama model is an auto-regressive language model, based on the transformer decoder architecture. To generate text, Llama processes a sequence of words as input and iteratively predicts the next token using a sliding window.
+LLaMA's architecture is slightly different from models like GPT-3. For instance, Llama employs the SwiGLU activation function rather than ReLU and opts for rotary positional embeddings in place of absolute learnable positional embeddings. 
```
Just to point out the consistency between LLaMA and Llama. Meta used LLaMA for the first version, and "Llama 2" for the next one. Maybe we should do the same, or just use "Llama" everywhere.

mehdiir replied:
Done.

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+ 
+The recent released Llama 2 introduced architectural refinements to better leverage very long sequences by extending the context length (up to 4096 tokens) and using grouped-query attention (GQA) decoding. 
+
+## [Mistral 7B](https://arxiv.org/abs/2310.06825)
```
```suggestion
### [Mistral 7B](https://arxiv.org/abs/2310.06825)
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+## [Mistral 7B](https://arxiv.org/abs/2310.06825)
+
+Mistral models, specifically Mistral 7B v0.1 is the first Large Language Models (LLMs) introduced by Mistral AI. The model contains 7.3 billion parameters. 
```
```suggestion
Mistral 7B v0.1, with 7.3 billion parameters, is the first LLM introduced by Mistral AI.
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+## [Mistral 7B](https://arxiv.org/abs/2310.06825)
+
+Mistral models, specifically Mistral 7B v0.1 is the first Large Language Models (LLMs) introduced by Mistral AI. The model contains 7.3 billion parameters. 
+The main innovative techniques used in of Mistral 7B's architecture are: 
```
```suggestion
The main novel techniques used in of Mistral 7B's architecture are: 
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+
+Mistral models, specifically Mistral 7B v0.1 is the first Large Language Models (LLMs) introduced by Mistral AI. The model contains 7.3 billion parameters. 
+The main innovative techniques used in of Mistral 7B's architecture are: 
+- Sliding Window Attention: Replace the full attention (square compute cost) with a sliding window based attention where each token can attend to at most 4,096 tokens from the previous layer (linear compute cost). This mechanism enables Mistral 7B to handle longer sequences, where higher layers can access to historical information beyond the window size of 4,096 tokens. 
```
```suggestion
- Sliding Window Attention: Replace the full attention (square compute cost) with a sliding window based attention where each token can attend to at most 4,096 tokens from the previous layer (linear compute cost). This mechanism enables Mistral 7B to handle longer sequences, where higher layers can access historical information beyond the window size of 4,096 tokens. 
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+Mistral models, specifically Mistral 7B v0.1 is the first Large Language Models (LLMs) introduced by Mistral AI. The model contains 7.3 billion parameters. 
+The main innovative techniques used in of Mistral 7B's architecture are: 
+- Sliding Window Attention: Replace the full attention (square compute cost) with a sliding window based attention where each token can attend to at most 4,096 tokens from the previous layer (linear compute cost). This mechanism enables Mistral 7B to handle longer sequences, where higher layers can access to historical information beyond the window size of 4,096 tokens. 
+- Grouped-query Attention: used in Llama 2 as well, the technique optimizes the inference process (reduce processing time) by caching the key and value vectors for previous decoded tokens in the sequence.  
```
```suggestion
- Grouped-query Attention: used in Llama 2 as well, the technique optimizes the inference process (reduce processing time) by caching the key and value vectors for previously decoded tokens in the sequence.  
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+- Sliding Window Attention: Replace the full attention (square compute cost) with a sliding window based attention where each token can attend to at most 4,096 tokens from the previous layer (linear compute cost). This mechanism enables Mistral 7B to handle longer sequences, where higher layers can access to historical information beyond the window size of 4,096 tokens. 
+- Grouped-query Attention: used in Llama 2 as well, the technique optimizes the inference process (reduce processing time) by caching the key and value vectors for previous decoded tokens in the sequence.  
+
+# PEFT
+
+PEFT, Parameter Efficient Fine-Tuning, is a collection of techniques (p-tuning, prefix-tuning, IA3, Adapters, and LoRa) designed to fine-tune large models using a much smalled set of training parameters while preserving the performance levels typically achieved through full fine-tuning. 
+
+
+
+## [LoRA](https://arxiv.org/abs/2106.09685)
+
+LoRA, Low-Rank Adaptation, shares similarities with Adapter layers. However, its primary objective is to achieve a more substantial reduction in the count of adjustable parameters. LoRA's operation involves the alteration of the training and updating of the modifiable parameters within the neural network. 
```
I'd maybe just use LoRA as the section header to avoid too much nesting here. We can still introduce PEFT inside the section.

mehdiir replied:
Done.

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,798 @@
+
+
+
+# Setup
+
+RoBERTa has a limitatiom of maximum sequence length of 512, so we set the `MAX_LEN=512` for all models to ensure a fair comparison.
+
+```python
+MAX_LEN = 512 
+roberta_checkpoint = "roberta-large"
+mistral_checkpoint = "mistralai/Mistral-7B-v0.1"
+llama_checkpoint = "meta-llama/Llama-2-7b-hf"
+```
+
+# Data preparation
+## A. Data loading
````
Same here, I'd maybe simplify the nesting structure and remove the short `Setup` section.

mehdiir replied:
The setup section is important as I define the checkpoints as maximum length.

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+import pandas as pd
+import os
+DATA_PATH = "~/peft_tweets/data"
+train_df=pd.read_csv(os.path.join(DATA_PATH, 'train.csv'))
+test_df=pd.read_csv(os.path.join(DATA_PATH, 'test.csv'))
```
```suggestion
train_df = pd.read_csv(os.path.join(DATA_PATH, 'train.csv'))
test_df = pd.read_csv(os.path.join(DATA_PATH, 'test.csv'))
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,798 @@
+
+```python
+from datasets import Dataset
+#Convert the training dataframe to HuggingFace dataset
````
I'd remove these comments as you have already described the tasks in the info text, they can be distracting while reading code. If you do decide to keep them, please use a space after the comment :)

mehdiir replied:
Done.

pcuenca reviewed (commented):
Great work! I did a first pass and suggested some format changes (there must be a single first-level `#` with the title, sections should be `##`, and so on).

mehdiir commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+peft==0.5.0
+scikit-learn==1.3.1
+torch==2.1.0
+transformers @ git+https://github.com/huggingface/transformers.git@21dc5859421cf0d7d82d374b10f533611745a8c5
```
Done. I unpined all packages and added the version in the wandb reports.

mehdiir commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+# Pre-trained Models
+## [RoBERTa](https://arxiv.org/abs/1907.11692)
+
+RoBERTa (Robustly Optimized BERT Approach) is an advanced variant of the BERT model proposed by Meta AI research team. It's a transformer-based language model using self-attention mechanisms for contextual word representations and trained with masked language model.
```
Done.

mehdiir commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+LoRA, Low-Rank Adaptation, shares similarities with Adapter layers. However, its primary objective is to achieve a more substantial reduction in the count of adjustable parameters. LoRA's operation involves the alteration of the training and updating of the modifiable parameters within the neural network. 
+
+![image](lora.png)
+
```
gaps removed.

mehdiir commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,798 @@
+```python
+import pandas as pd
+import os
+DATA_PATH = "~/peft_tweets/data"
````
That's a great point, I am currently working on it.

mehdiir replied:
@merveenoyan, thank you for the suggestion. I pushed the dataset to HF and updated the data section accordingly.

mehdiir commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+import os
+DATA_PATH = "~/peft_tweets/data"
+train_df=pd.read_csv(os.path.join(DATA_PATH, 'train.csv'))
+test_df=pd.read_csv(os.path.join(DATA_PATH, 'test.csv'))
```
Working on storing the data in HF datasets.

mehdiir commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,798 @@
+# Table of contents
+<!-- TOC -->
+
+- [Table of contents](#table-of-contents)
```
Removed.

mehdiir commented on _blog.yml:
```diff
@@ -2968,4 +2968,16 @@
+  title: "Comparing the Performance of LLMs: A Deep Dive into Roberta, Llama, and Mistral for Disaster Tweets Analysis with Lora"
+  author: mehdiiraqui
+  thumbnail: /blog/assets/Lora-for-sequence-classification-with-Roberta-Llama-Mistral/Thumbnail.png
+  date: October 18, 2023
```
Updated.

mehdiir commented:
> Great work! I did a first pass and suggested some format changes (there must be a single first-level `#` with the title, sections should be `##`, and so on).

@pcuenca  Thank you for your review. I have addressed all the suggestions. Please let me know if the new version looks good to you.

mehdiir commented:
> Very interesting findings, I was wondering how larger autoregressive models would compare with encoder only models. I left a couple of comments about the code and did nitpicking, thanks a lot for this!

@merveenoyan Thank you for your review. Indeed, the results consolidate the observation that encoder models are more suitable for NLU tasks than autoregressive ones. However, it could be interesting to investigate whether prompt tuning could enhance the performance of autoregressive models in understanding the classification task (maybe in a future blogpost 😄).

Also, I have addressed all the suggestions. Please let me know if the new version looks good to you.

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+```
+
+
+You can follow the same steps for preparing the data for Mistral 7B and Lllama 2 models: 
````
```suggestion
You can follow the same steps for preparing the data for Mistral 7B and Llama 2 models: 
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+
+```python
+# Load Mistral 7B Tokenizer
+# Add prefix space to tokenize words into subwords
````
Why do we need this?

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,792 @@
+# Load Llama 2 Tokenizer
+# Add prefix space to tokenize words into subwords
+from transformers import AutoTokenizer, DataCollatorWithPadding
+llama_tokenizer = AutoTokenizer.from_pretrained(llama_checkpoint,  add_prefix_space=True)
```
```suggestion
llama_tokenizer = AutoTokenizer.from_pretrained(llama_checkpoint, add_prefix_space=True)
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+Then, we compute the maximum length of the column text:
+```python
+# Number of Characters
+max_char=data['train'].to_pandas()['text'].str.len().max()
````
```suggestion
max_char = data['train'].to_pandas()['text'].str.len().max()
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+```
+
+```
+The maximum number of character is 152.
+The maximum number of word is 31.
````
Why do we need this information? We are later padding each batch.

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+
+```python
+from transformers import AutoModelForSequenceClassification 
+# Load a pre-trained model with a sequence classification header 
````
```suggestion
```

It's exactly what's explained in the previous sentence.

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+```
+
+
+####  LoRa setup for RoBERTa classifier
````
```suggestion
####  LoRA setup for RoBERTa classifier
```

Could we use it consistently elsewhere, if possible?

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,792 @@
+- lora_dropout: Dropout probability of the LoRA layers
+- bias: Whether to add bias term to LoRa layers
+
+In the code below, we use the values recommended by the [Lora paper](https://arxiv.org/abs/2106.09685).
+
+Note that in this blog, we will perform a hyperparameter tuning with wandb.
```
```suggestion
The code below uses the values recommended by the [Lora paper](https://arxiv.org/abs/2106.09685). [Later in this post](#hyperparameter-tuning) we will perform hyperparameter tuning of these parameters using `wandb`.
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+roberta_model.print_trainable_parameters()
+```
+
+We can see below that the number of trainable parameters reprents only 0.64% of the RoBERTa model parameters:
````
```suggestion
We can see that the number of trainable parameters represents only 0.64% of the RoBERTa model parameters:
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+Let's load the pre-trained Mistral-7B model with a sequence classification head:
+
+```python
+from transformers import AutoModelForSequenceClassification # Load a pre-trained model with a sequence classification header 
````
```suggestion
from transformers import AutoModelForSequenceClassification
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+Let's load pre-trained Llama 2 model with a sequence classification header. 
+
+```python
+from transformers import AutoModelForSequenceClassification # Load a pre-trained model with a sequence classification header 
````
```suggestion
from transformers import AutoModelForSequenceClassification
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+First, we define the performance metrics we will use to compare the three models: F1 score, recall, precision and accuracy:
+
+```python
+#Define evaluation metrics
````
```suggestion
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,792 @@
+#Define evaluation metrics
+import evaluate
+import numpy as np
+def compute_metrics(eval_pred):
```
```suggestion

def compute_metrics(eval_pred):
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+```
+
+### Custom Trainer for Weighted Loss 
+As mentioned at the beginning of this post, we have an imbalanced distribution between positive and negative classes. We need to train our models with a weight cross-entropy loss to account for that. The `Trainer` class doesn't support providing a custom loss as it expects to get the loss directly from the model's outputs. 
````
```suggestion
As mentioned at the beginning of this post, we have an imbalanced distribution between positive and negative classes. We need to train our models with a weighted cross-entropy loss to account for that. The `Trainer` class doesn't support providing a custom loss as it expects to get the loss directly from the model's outputs. 
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+
+```python
+from transformers import Trainer
+class WeightedCELossTrainer(Trainer):
````
```suggestion

class WeightedCELossTrainer(Trainer):
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+```python
+from transformers import TrainingArguments, Trainer
+
+mistral_model = mistral_model.cuda()
````
```suggestion
llama_model = llama_model.cuda()
```

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+
+
+
+```python
+lr = 2e-4
+batch_size = 8
+num_epochs = 5
+
+training_args = TrainingArguments(
+    output_dir="llama-lora-token-classification",
+    learning_rate=lr,
+    lr_scheduler_type= "constant",
+    warmup_ratio= 0.1,
+    max_grad_norm= 0.3,
+    per_device_train_batch_size=batch_size,
+    per_device_eval_batch_size=batch_size,
+    num_train_epochs=num_epochs,
+    weight_decay=0.001,
+    evaluation_strategy="epoch",
+    save_strategy="epoch",
+    load_best_model_at_end=True,
+    report_to="wandb",
+    fp16=True,
+    gradient_checkpointing=True,
+)
+
+```
+Then, we define the Llama 2 trainer:
+
+```python
+llama_trainer = WeightedCELossTrainer(
+    model=llama_model,
+    args=training_args,
+    train_dataset=llama_tokenized_datasets['train'],
+    eval_dataset=llama_tokenized_datasets["val"],
+    data_collator=llama_data_collator,
+    compute_metrics=compute_metrics
+)
+```
````
This block appears to be repeated except the `lr` line.

mehdiir replied:
Thanks for the catch! I removed all the duplicated codes.

pcuenca commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,792 @@
+## Hardware Used 
+
+- Number of nodes: 1 
+- Number of GPUs per node: 2
```
Are we using the 2 nodes? It looks like each training job fits fine in just one card, is that correct?

mehdiir replied:
I am using a local workstation with 2 GPUs, so it's one node with 2 GPUs

pcuenca replied:
Sorry I meant the GPUs 🤦 If only one is required I wouldn't mention we are using 2, so more people can be encouraged to try this out. One of the advantages of LoRA is that it doesn't require as many resources as regular fine-tuning.

mehdiir replied:
I run a test with 1 GPU and was able to run all the 3 models successfully. I changed the number of GPUs to 1.

pcuenca reviewed (commented):
Thanks a lot for iterating!

mehdiir commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,792 @@
+
+LoRA, Low-Rank Adaptation, is a PEFT method that shares similarities with Adapter layers. However, its primary objective is to reduce the model's trainable parameters. LoRA's operation involves the alteration of the training and updating of the modifiable parameters within the neural network. 
+
+![image](lora.png)
```
I have created a PR :https://huggingface.co/datasets/huggingface/documentation-images/discussions/203
Seems still open.

pcuenca replied:
Merged. You need to use this URL:

![LoRA diagram](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/Lora-for-sequence-classification-with-Roberta-Llama-Mistral/lora.png).

mehdiir commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,792 @@
+```python
+col_to_delete = ['id', 'keyword','location', 'text']
+# Apply the preprocessing function
+roberta_tokenized_datasets = data.map(roberta_preprocessing_function, batched=False)
````
My bad, I used that for debugging and completely forgot to remove it from the final version. 
I just changed it

mehdiir commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,792 @@
+# Apply the preprocessing function
+roberta_tokenized_datasets = data.map(roberta_preprocessing_function, batched=False)
+#Remove the undesired columns
+roberta_tokenized_datasets = roberta_tokenized_datasets.remove_columns(col_to_delete)
```
The RoBERTa tokenizer has been trained to treat spaces as part of the tokens, so the first word of the sentence is encoded differently. Hugging Face's documentation recommends setting 'add_prefix_space' to 'True' to ensure that the first word is encoded correctly. To ensure consistent pre-processing for all three models, I have set the parameters to 'True'. I added the a note to the post.

mehdiir commented:
> Thanks a lot for iterating!

@pcuenca Thanks again for you reviews. I addressed all of them.

pcuenca reviewed (approved):
Thanks again @mehdiir for addressing the comments! Looks good to me. @merveenoyan  can you maybe take a quick look to see if it's ok for you?

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,746 @@
+# Comparing the Performance of LLMs: A Deep Dive into Roberta, Llama 2, and Mistral for Disaster Tweets Analysis with Lora
+<!-- TOC -->
+
+- [Comparing the Performance of LLMs: A Deep Dive into Roberta, Llama 2, and Mistral for Disaster Tweets Analysis with Lora](#comparing-the-performance-of-llms-a-deep-dive-into-roberta-llama-2-and-mistral-for-disaster-tweets-analysis-with-lora)
```
```suggestion
- [Comparing the Performance of LLMs: A Deep Dive into Roberta, Llama 2, and Mistral for Disaster Tweets Analysis with LoRA](#comparing-the-performance-of-llms-a-deep-dive-into-roberta-llama-2-and-mistral-for-disaster-tweets-analysis-with-lora)
```

merveenoyan replied:
```suggestion
- [Comparing the Performance of LLMs: A Deep Dive into Roberta, LLaMa 2, and Mistral for Disaster Tweets Analysis with Lora](#comparing-the-performance-of-llms-a-deep-dive-into-roberta-llama-2-and-mistral-for-disaster-tweets-analysis-with-lora)
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,746 @@
+    - [Models](#models)
+        - [RoBERTa](#roberta)
+            - [Load RoBERTA Checkpoints for the Classification Task](#load-roberta-checkpoints-for-the-classification-task)
+            - [LoRa setup for RoBERTa classifier](#lora-setup-for-roberta-classifier)
```
```suggestion
            - [LoRA setup for RoBERTa classifier](#lora-setup-for-roberta-classifier)
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,746 @@
+            - [LoRa setup for RoBERTa classifier](#lora-setup-for-roberta-classifier)
+        - [Mistral](#mistral)
+            - [Load checkpoints for the classfication model](#load-checkpoints-for-the-classfication-model)
+            - [LoRa setup for Mistral 7B classifier](#lora-setup-for-mistral-7b-classifier)
```
```suggestion
            - [LoRA setup for Mistral 7B classifier](#lora-setup-for-mistral-7b-classifier)
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,746 @@
+        - [Mistral](#mistral)
+            - [Load checkpoints for the classfication model](#load-checkpoints-for-the-classfication-model)
+            - [LoRa setup for Mistral 7B classifier](#lora-setup-for-mistral-7b-classifier)
+        - [Llama 2](#llama-2)
```
```suggestion
        - [LLaMa 2](#llama-2)
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,746 @@
+            - [LoRa setup for Mistral 7B classifier](#lora-setup-for-mistral-7b-classifier)
+        - [Llama 2](#llama-2)
+            - [Load checkpoints for the classfication mode](#load-checkpoints-for-the-classfication-mode)
+            - [LoRa setup for Llama 2 classifier](#lora-setup-for-llama-2-classifier)
```
```suggestion
            - [LoRA setup for LLaMa 2 classifier](#lora-setup-for-llama-2-classifier)
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,746 @@
+            - [Load checkpoints for the classfication model](#load-checkpoints-for-the-classfication-model)
+            - [LoRa setup for Mistral 7B classifier](#lora-setup-for-mistral-7b-classifier)
+        - [Llama 2](#llama-2)
+            - [Load checkpoints for the classfication mode](#load-checkpoints-for-the-classfication-mode)
```
```suggestion
            - [Load checkpoints for the classification mode](#load-checkpoints-for-the-classfication-mode)
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,746 @@
+        - [Trainer Setup](#trainer-setup)
+            - [RoBERTa](#roberta)
+            - [Mistral-7B](#mistral-7b)
+            - [Llama 2](#llama-2)
```
```suggestion
            - [LLaMa 2](#llama-2)
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
```diff
@@ -0,0 +1,746 @@
+
+In this blog, we used PEFT (Parameter-Efficient Fine-Tuning) technique: LoRA (Low-Rank Adaptation of Large Language Models) for fine-tuning the pre-trained model on the sequence classification task. LoRa is designed to significantly reduce the number of trainable parameters while maintaining strong downstream task performance. 
+
+The main objective of this blogpost is to implement LoRA fine-tuning for sequence classification task using three pre-trained models from Hugging Face: [meta-llama/Llama-2-7b-hf](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf), [mistralai/Mistral-7B-v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1), and [roberta-large](https://huggingface.co/roberta-large)
```
```suggestion
The main objective of this blog post is to implement LoRA fine-tuning for sequence classification tasks using three pre-trained models from Hugging Face: [meta-llama/Llama-2-7b-hf](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf), [mistralai/Mistral-7B-v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1), and [roberta-large](https://huggingface.co/roberta-large)
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,746 @@
+
+## Dependencies
+
+```python
````
```suggestion
```bash
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,746 @@
+
+Here's an overview of the dataset:
+
+```
````
```suggestion
```bash
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,746 @@
+    return roberta_tokenizer(examples['text'], truncation=True, max_length=MAX_LEN)
+```
+
+By applying the preprocessing function to the first example of our training dataset, we can see that we have inputs ids and an attention mask:
````
```suggestion
By applying the preprocessing function to the first example of our training dataset, we have the tokenized inputs (`input_ids`) and the attention mask:
```

merveenoyan commented on Lora-for-sequence-classification-with-Roberta-Llama-Mistral.md:
````diff
@@ -0,0 +1,746 @@
+```
+
+We can see that the number of trainable parameters represents only 0.64% of the RoBERTa model parameters:
+```
````
```suggestion
```bash
```

merveenoyan reviewed (commented):
Thank you! I added a bit of a clarification and did nit-picking. Could you please reflect the suggestions I made to rest of the text for consistency?

mehdiir commented:
@merveenoyan Llama 2 refers to the model name, but LlaMa is the LLM.

mehdiir commented:
> Thank you! I added a bit of a clarification and did nit-picking. Could you please reflect the suggestions I made to rest of the text for consistency?

Thanks @merveenoyan  for the review! I included all the suggestions. One clarification: I kept "Llama 2" instead of "LlaMa 2" because META uses `LlaMa` to refer to the first published LLM and `Llama 2` for the recent one.

pcuenca commented:
@merveenoyan feel free to merge when you are ready :)

pcuenca commented:
Congrats @mehdiir for the great work, and thanks a lot @merveenoyan for your thorough reviews 🎉
